optimization-cookbook.md (WIP) (#3409)
circleci/circleci-docs: jekyll/_cci2/optimization-cookbook.md

---
layout: classic-docs
title: "CircleCI Optimizations Cookbook"
short-title: "Optimizations Cookbook"
description: "Starting point for Optimizations Cookbook"
categories: [getting-started]
order: 1
---

> The *CircleCI Optimizations Cookbook* is a collection of individual use cases (referred to as "recipes") that provide you with detailed, step-by-step instructions on how to perform various optimization tasks using CircleCI resources. This guide, and it associated sections, will enable you to quickly and easily perform easily repeatable optimization tasks on the CircleCI platform.

Reviewer: Who is this doc for, and why should I read it?

What are the concrete use-cases I will get out of this document. An example of this would be, reducing build minutes or developer time. And how.

My usual thought pattern when looking through examples or cookbooks, recipes, etc; is asking "What problem does this solve? And is that a problem I have now?"

An example of this might be; "OOoo, a recipe for baked potatoes; hmmm I have a problem in that we bought too many potatoes! Perfect."

Author: Rewrote introduction to clarify the purpose of this guide.

Author: Fixed in document.

* TOC
{:toc}

## Introduction

> When work with the CircleCI platform, one of the most critical tasks you will want to complete is to optimize your jobs, builds, workflows, and other tasks to prevent performance degradation, latency, and concurrency. Optimizing your tasks is not simply a function of lessening the workload or number of tasks, but rather, reviewing your optimization and caching strategies to ensure you are meeting your business and technical requirements while also maintaining stability and reliability.

Reviewer 2: I agree.

My approach here, would be after getting things working on CircleCI, optimizing comes next because 1) You are a business and decreasing minutes/resources means savings or 2) For anyone, less complex, more efficient builds means saving time.

> When work with the CircleCI platform, one of the most critical tasks you will want to complete is to optimize your jobs, builds, workflows, and other tasks to prevent performance degradation, latency, and concurrency. Optimizing your tasks is not simply a function of lessening the workload or number of tasks, but rather, reviewing your optimization and caching strategies to ensure you are meeting your business and technical requirements while also maintaining stability and reliability.

Author: fixed in documentation.

CircleCI has developed several different approaches you may wish to use to better optimize these tasks while also leveraging the power of the CircleCI platform and its associated resources (e.g. orbs). The sections below describe various strategies and methodologies you can use to optimize your organization tasks.

### Optimization Recipes

The table below lists the different optimization tasks that can be performed using the CircleCI platform.

Recipe | Description 
------------|-----------
Single-Threading Builds (Queueing) | This section describes how you can use a specific single-threading (queueing) orb to ensure proper job and build completion without concurrency (queueing subsequent builds or jobs).
Caching |  This section describes how you can use different caching strategies in your implemenations to optimize your builds and workflows.
Workflow Optimizations | This section describes how you can optimize your workflows.

> ## Single-Threading Builds (Queuing)

Reviewer 3: I'm not clear about the use-case for this.

I thought the queue orb was specifically about managing access to an external shared resource?

Reviewer: +1, I'm also unsure the application of "thread" in this case makes sense. Given our system, I would consider parallel containers to be more like threads than parallel builds (for example). Referencing terms that have explicit meaning when applied ambiguously will leave it up to the reader to determine the meaning which can lead to misinterpretation.

Reviewer 4: Running Jobs Sequentially?
Forcing Jobs to Run Sequentially?

Author: Re-titled section to "Running Jobs Sequentially" and made some content changes to address these wording issues.

> One of the most common tasks you may encounter when using the CircleCI platform for builds and jobs is managing multiple jobs and builds simultaneously to ensure your operations do not fail because of system timeouts. This becomes especially important when you have multiple contributors and committers working in the same environment. Because the CircleCI platform was designed to handle multiple tasks simultaneously without performance degradation or latency, concurrency may sometimes become an issue if there are a large number of jobs being queued, waiting for a previous job to be completed before the new job can be initiated, and the system timeout is set too low. In this case, one job will be completed, and other jobs may fail due to this timeout setting.

Reviewer 4: I feel like we need to be careful using the terms job and build like this - my understanding 'build' is kind of a generic terms that has many subtle meanings inside and outside circleci, whereas job is a very specific term, e.g. you can have a build job. 

Here, for example, could just say: "One of the most common tasks you may encounter when using the CircleCI platform is managing multiple jobs simultaneously to ensure your operations do not fail because of system timeouts." and this has the same meaning but is less ambiguous?

That being said - I'm struggling to understand this use case - why would we want to prevent a job from running concurrently with others?

Reviewer 4: Is system timeout something the customer has control over?

Author: Resolved in document. Thanks for the great comments @rosieyohannan 👍

To better optimize builds and jobs and prevent concurrency and subsequent jobs failing because of timeout, CircleCI has developed a single-threading (queueing) orb that specifically addresses these performance issues. By invoking this orb, you can greatly improve overall job and build performance and prevent concurrency.

> **Note**For more detailed information about the CircleCI Queueing orb, refer to the following CircleCI pages:

Reviewer 5: ```suggestion
**Note:** For more detailed information about the CircleCI Queueing orb, refer to the following CircleCI pages:
```
Just add a space between `Note` and the word `For`

- Queueing and Single Threading Overview - https://github.com/eddiewebb/circleci-queue
- CircleCI Queueing Orb - https://circleci.com/orbs/registry/orb/eddiewebb/queue#quick-start

### Setting Up and Configuring Your Environment to use the CircleCI Platform and CircleCI Orbs

To configure your environment for the CircleCI platform and CircleCI orbs, follow the steps listed below.

1) Use CircleCI `version 2.1` at the top of your `.circleci/config.yml` file.

`version: 2.1`

2) If you do not already have Pipelines enabled, you'll need to go to **Project Settings -> Advanced Settings** to enable pipelines.

3) Add the `orbs` stanza below your version, invoking the orb:

`orbs:
  queue: eddiewebb/queue@1.1.2`

4) Use `queue` elements in your existing workflows and jobs.

> 5) Opt-in to use of third-party orbs on your organization’s **Security Settings** page.

Reviewer 3: If we're recommending it in docs, we should probably make it an orb that doesn't need to be opt-ed in?

Reviewer 2: I agree, though I personally don't know anything about this orb and its use case.

Author: fixed

Reviewer 4: Does this opt in step need to come before step 4?

### Blocking Workflows

> One of the easiest ways to prevent workflow concurrency using the CircleCI Single-Threading orb is to enable "blocking" of any workflows with an earlier timestamp. By setting the `block-workflow` parameter value to `true`, all workflows will be forced to run consecutively, not concurrently, thereby limiting the number of workflows in the queue. In turn, this will also improve overall performance while making sure no workflows are discarded.

Reviewer 4: What would cause a workflow to be discarded?

Author: Re-worded this section to make it more clear about workflow blocking and concurrency.

    description: >-
      If true, this job will block until no other workflows with an earlier
      timestamp are running. Typically used as first job.
    type: boolean
  consider-branch:
    default: true
    description: Should we only consider jobs running on the same branch?
    type: boolean
  consider-job:
    default: false
    description: Deprecated. Please see block-workflow.
    type: boolean
  dont-quit:
    default: false
    description: >-
      Quitting is for losers. Force job through once time expires instead of
      failing.
    type: boolean
  only-on-branch:
    default: '*'
    description: Only queue on specified branch
    type: string
  time:
    default: '10'
    description: How long to wait before giving up.
    type: string
  vcs-type:
    default: github
    description: Override VCS to 'bitbucket' if needed.
    type: string
steps:
  - until_front_of_line:
      consider-branch: <<parameters.consider-branch>>
      consider-job: <<parameters.consider-job>>
      dont-quit: <<parameters.dont-quit>>
      only-on-branch: <<parameters.only-on-branch>>
      time: <<parameters.time>>
      vcs-type: <<parameters.vcs-type>>
```
{% endraw %}

> ## Caching

Reviewer: Caching what exactly? "Caching dependencies" might be a more suitable title for this section. IMO the more specific the recipe the better. "Fried Brussel sprouts" vs "Fried food": which one sounds more appetizing/relevant to you?

Author: fixed in document.

One of the quickest and easiest ways to optimize your builds and workflows is to implement specific caching strategies so you can use existing data from previous builds and workflows. Whether you choose to use a package management application (e.g. Yarn, Bundler, etc), or manually configure your caching, utilizing the best and most effective caching strategy may improve overall performance. In this section, several different use cases are described that may assist you in determining which caching method is best for your implementation.

> If a job fetches data at any point, it is likely that you can make use of caching. A common example is the use of a package/dependency manager. If your project uses Yarn, Bundler, or Pip, for example, the dependencies downloaded during a job can be cached for later use rather than being re-downloaded on every build. The example below shows how you can use a package manager for caching.

Reviewer 2: > The example below shows how you can use a package manager for caching.

This sentence should maybe be reversed?

"The example below shows how you can use caching for a package manager."

Author: Fixed...thanks for the help on this one.

  build:
    steps: # a collection of executable commands making up the 'build' job
      - checkout # pulls source code to the working directory
      - restore_cache: # **restores saved dependency cache if the Branch key template or requirements.txt files have not changed since the previous run**
          key: deps1-{{ .Branch }}-{{ checksum "requirements.txt" }}
      - run: # install and activate virtual environment with pip
          command: |
            python3 -m venv venv
            . venv/bin/activate
            pip install -r requirements.txt
      - save_cache: # ** special step to save dependency cache **
          key: deps1-{{ .Branch }}-{{ checksum "requirements.txt" }}
          paths:
            - "venv"
```
{% endraw %}

Notice in the above example that you can use a `checksum` in the cache key. This is used to calculate when a specific dependency-management file (such as a `package.json` or `requirements.txt` in this case) changes and so the cache will be updated accordingly. Also note that the `restore_cache` example uses interpolation to put dynamic values into the cache-key, allowing more control in what exactly constitutes the need to update a cache.

> **Note:** Before adding any chaching steps to your workflow, verify the dependencies installation step succeeds. Caching a failed dependency step will require you to change the cache key in order to avoid failed builds due to a bad cache. 

Reviewer 2: > chaching

Author: Fixed.

**Note:** Before adding any caching steps to your workflow, verify the dependencies installation step succeeds. Caching a failed dependency step will require you to change the cache key in order to avoid failed builds due to a bad cache. 

Because caching is a such a critical aspect of optimizing builds and workflows, you should first familiarize yourself with the following page that describes caching and how various strategies can be optimized:

- Caching - https://circleci.com/docs/2.0/caching/

### Caching Example

<Example needed here>

## Testing Optimization

When running tests on the CircleCI platform, one of the primary considerations you will want to make is how you can optimize the testing process to minimize credit usage and improve overall testing performance and results. Testing can sometimes be a time and performance-intensive process; therefore, the ability to reduce testing time can be a significant boost to your organizational goals.

> There are many different test suites and approaches you can use when testing on the CircleCI platform. Although CircleCI is test suite agnostic, the example below illustrates an example of how you could optimize testing using Django and the CircleCI platform.

Reviewer 6: I humbly suggest removing the last of the "example"s in that last sentence:
> the example below illustrates ~an example of~ how

Author: Thanks for the good catch. Fixed in document.

### Testing Optmization on the CircleCI Platform For a Python Django Project

> 

Reviewer 6: This section uses significant portions of my [blog post](https://www.brautaset.org/articles/2019/speed-up-circleci.html) without attribution. I would appreciate if you could work "adapted (with permission) from [this blog post](https://www.brautaset.org/articles/2019/speed-up-circleci.html)" into the text.

Author: @stig . Absolutely will make a note of this in the document so there is proper attribution. Thanks for the help on this :)

> Some organizations use CircleCI to run tests for each change before merging to the main branch. Faster tests means faster feedback cycles, which in turn means you can confidently ship code more often. For example, if you perform testing for a Python Django's application workflow, it might take as long as 13-15 minutes to complete the testing on the CircleCI platform. Most organizations would deem this testing to be too expensive and look at ways to reduce this time to preserve credit usage and improve overall testing performance.

Author: Thanks for the comment. Much better and more clear wording. Fixed in document.

- Persisting the image to the workspace added another **60** seconds.
- The test jobs then had to attach the workspace and load the base image, adding another **30** seconds to the testing.
- The test jobs started dependencies (Redis, Cassandra, and PostgreSQL) with `docker-compose`, which required the use of the machine executor. This added an additional **30-60** second startup time compared to the docker executors.
- Because the base image from the build job contained only runtime dependencies, a docker image had to be built, extending the base to add dependencies for testing. This added  about **70** seconds.

As you can see, there is a a significant amount of time being spent setting up the tests, without any actual tests being performed. In fact, this approach required 6.5 minutes before the actual tests were run, which took another 6.5 minutes.

### Test Preparation Optimization

Knowing that ~13 minutes was too long to perform the steps in this workflow, the following approaches were taken to optimize this workflow to reduce the time it took to run this workflow.

#### Changing the CI Test Workflow

> The CI test workflow was changed to no longer depend on building the base image. The test jobs were also changed to lauch auxiliary services using CircleCI's docker executor native service container support instead of using `docker-compose`. Finally,`tox` was run from the main container to install dependencies and run tests, which eliminates minutes used to save and then restore the image from the workspace. This also eliminated the extra start-up costs of the machine executor.

Author: Fixed.

#### Dependency Changes

> Installing dependencies in the primary container on CircleCI, rather than relying on a Dockerfile, may enable you to use CircleCI's caching to speed up `virtualenv` creation. Even if you do not have access to this shared cache when installing dependencies, you may save around **90** seconds when building the virtualenvs.

Reviewer 6: > Even if you do not have access to this shared cache when installing dependencies, you may save around **90** seconds when building the virtualenvs.

I'm afraid I don't follow. How can you save 90 seconds when building the virtualenvs "even if you do not have access to the shared cache"?

Author: @stig You're right, this was confusing, so I just went ahead and removed this from the page. Thanks for the good catch.

### Test Execution Optimization

Now that the test preparation time has been reduced, you may also wish to speed up the running of the actual tests. For example, you may not need to keep the database after test runs. One way you could speed up testing is to replace the database image used for tests with an in-memory Postgres image that does not save to disk. Another method you may wish to take is to run your tests in parallel instead of one-test-at-a-time. 

The figure below illustrates how overall these changes can reduce the total workflow time.

![Test Optimization Process After Optimization]({{site.baseurl}}/assets/img/docs/optimization_cookbook_workflow_optimization_2.png)

> As you can see, there was no single step performed to reduce overall workflow time. For example, running tests in parallel would not have seen much benefit when most of the time was being used to prepare to the run the tests. By recognizing the differences between running tests on the CircleCI platform instead of a local context, and making a few changes to test preparation and execution, you may be able to see improved test run time. 

Reviewer 6: > prepare to ~the~ run the tests

Author: Fixed.

---
layout: classic-docs
title: "CircleCI Optimizations Cookbook"
short-title: "Optimizations Cookbook"
description: "Starting point for Optimizations Cookbook"
categories: [getting-started]
order: 1
---

> The *CircleCI Optimizations Cookbook* is a collection of individual use cases (referred to as "recipes") that provide you with detailed, step-by-step instructions on how to perform various optimization tasks using CircleCI resources. This guide, and it associated sections, will enable you to quickly and easily perform easily repeatable optimization tasks on the CircleCI platform.

Reviewer 4: ```suggestion
The *CircleCI Optimizations Cookbook* is a collection of individual use cases (referred to as "recipes") that provide you with detailed, step-by-step instructions on how to perform various optimization tasks using CircleCI resources. This guide, and it associated sections, will enable you to quickly and easily perform repeatable optimization tasks on the CircleCI platform.
```

* TOC
{:toc}

## Introduction

> Sometimes when you are using the CircleCI platform, you may encounter unexpected lags in pipeline performance, which can negatively affect your ability to perform critical organizational functions. These performance bottlenecks can not only impact overall performance, but also cause workflow and build failures. These "hiccups" can cost you money in terms of credit usage, resources, and individual time spent reducing these bottlenecks.

Reviewer 4: ```suggestion
Sometimes when you are using the CircleCI platform, you may encounter unexpected lags in pipeline performance, which can negatively affect your ability to perform critical organizational functions. These performance bottlenecks can not only impact overall performance, but also cause workflow and build failures. These "hiccups" can cost you money in terms of credit usage, resources, and individual time spent reducing bottlenecks.
```

> Sometimes when you are using the CircleCI platform, you may encounter unexpected lags in pipeline performance, which can negatively affect your ability to perform critical organizational functions. These performance bottlenecks can not only impact overall performance, but also cause workflow and build failures. These "hiccups" can end cost you money in terms of credit usage, resources, and individual time spent reducing these bottlenecks.

Reviewer 4: ```suggestion
Sometimes when you are using the CircleCI platform, you may encounter unexpected lags in pipeline performance, which can negatively affect your ability to perform critical organizational functions. These performance bottlenecks can not only impact overall performance, but also cause workflow and build failures. These "hiccups" can cost you money in terms of credit usage, resources, and individual time spent reducing these bottlenecks.
```

This guide provides you with the following optimization strategies that you can utilize to minimize any potential performance bottlenecks and ensure that you are getting the best performance possible when using CircleCI:

- Running jobs sequentially to prevent concurrency
- Implementing caching strategies to optimize builds and workflows

> - Improving text performance

Reviewer 4: ```suggestion
- Improving test performance
```

**Note:** This guide will be updated with new optimization strategies on a continual basis, so please feel free to refer to this page for new and updated content.

### Optimization Recipes

The table below lists the different optimization tasks that can be performed using the CircleCI platform.

> Recipe | Description 

Reviewer 4: This table is basically the same info as the bullet list above? maybe each can link to the section headings below?

Author: great comment @rosieyohannan 👍 I went ahead and removed the table and added a TOC so the user can easily jump to the different sections on the page. Thanks :)

------------|-----------
Running Jobs Sequentially To Prevent Concurrency | This section describes how you can use the queue orb to run jobs sequentially without concurrency.

> Using Caching to Optimizae Builds and Workflows |  This section describes how you can use different caching strategies in your implemenations to optimize your builds and workflows.

Reviewer 4: ```suggestion
Using Caching to Optimize Builds and Workflows |  This section describes how you can use different caching strategies in your implementations to optimize your builds and workflows.
```

Improving Test Performance | This section describes a specific use case where test performance was significantly improved on the CircleCI platform.

## Running Jobs Sequentially To Prevent Concurrency

> One of the most common tasks you may encounter when using the CircleCI platform is managing multiple jobs simultaneously to ensure your operations do not fail because of system timeouts. This becomes especially important when you have multiple contributors and committers working in the same environment. Because the CircleCI platform was designed to handle multiple tasks simultaneously without performance degradation or latency, concurrency may sometimes become an issue if there are a large number of jobs being queued, waiting for a previous job to be completed before the new job can be initiated, and the user has set the system timeout is too low. In this case, one job will be completed, and other jobs will fail due to this timeout setting.

Reviewer 4: ```suggestion
One of the most common tasks you may encounter when using the CircleCI platform is managing multiple jobs simultaneously to ensure your operations do not fail because of system timeouts. This becomes especially important when you have multiple contributors and committers working in the same environment. Because the CircleCI platform was designed to handle multiple tasks simultaneously without performance degradation or latency, concurrency may sometimes become an issue if there are a large number of jobs being queued, waiting for previous jobs to be completed before new jobs can be initiated, and system timeout is set too low. In this case, one job will be completed, and other jobs will fail due to this timeout setting.
```

> To better optimize builds and jobs and prevent concurrency and subsequent jobs failing because of timeout, CircleCI has developed a single-threading (queueing) orb that specifically addresses these performance issues. By invoking this orb, you can greatly improve overall job and build performance and prevent concurrency.

Reviewer 4: Still unclear about the use of terms 'build' and 'job' - I think workflows and jobs are what is being talked about?

Author: made this section more clear :)

**Note:** For more detailed information about the CircleCI Queueing orb, refer to the following CircleCI pages:

- Queueing and Single Threading Overview - https://github.com/eddiewebb/circleci-queue
- CircleCI Queueing Orb - https://circleci.com/orbs/registry/orb/eddiewebb/queue#quick-start

### Setting Up and Configuring Your Environment to use the CircleCI Platform and CircleCI Orbs

To configure the environment for the CircleCI platform and CircleCI orbs, follow the steps listed below.

1) Use CircleCI `version 2.1` at the top of your `.circleci/config.yml` file.

`version: 2.1`

2) If you do not already have Pipelines enabled, you'll need to go to **Project Settings -> Advanced Settings** to enable pipelines.

3) Add the `orbs` stanza below your version, invoking the orb. For example:

`orbs:
  queue: eddiewebb/queue@1.1.2`

> 4) Use `queue` elements in your existing workflows and jobs.

Reviewer 4: I would want some specifics on using the various elements in the recipe maybe?

Author: I will work with Eddie on getting some more specifics for this recipe for an iteration on this content in the next few days; however, I am going to publish the page so we can start getting user feedback.

    description: >-
      If true, this job will block until no other workflows with an earlier
      timestamp are running. Typically used as first job.
    type: boolean
  consider-branch:
    default: true
    description: Should we only consider jobs running on the same branch?
    type: boolean
  consider-job:
    default: false
    description: Deprecated. Please see block-workflow.
    type: boolean
  dont-quit:
    default: false
    description: >-
      Quitting is for losers. Force job through once time expires instead of
      failing.
    type: boolean
  only-on-branch:
    default: '*'
    description: Only queue on specified branch
    type: string
  time:
    default: '10'
    description: How long to wait before giving up.
    type: string
  vcs-type:
    default: github
    description: Override VCS to 'bitbucket' if needed.
    type: string
steps:
  - until_front_of_line:
      consider-branch: <<parameters.consider-branch>>
      consider-job: <<parameters.consider-job>>
      dont-quit: <<parameters.dont-quit>>
      only-on-branch: <<parameters.only-on-branch>>
      time: <<parameters.time>>
      vcs-type: <<parameters.vcs-type>>
```
{% endraw %}

> ## Using Caching to Optimizae Builds and Workflows

Reviewer 4: ```suggestion
## Using Caching to Optimize Builds and Workflows
```

      - save_cache: # ** special step to save dependency cache **
          key: deps1-{{ .Branch }}-{{ checksum "requirements.txt" }}
          paths:
            - "venv"
```
{% endraw %}

Notice in the above example that you can use a `checksum` in the cache key. This is used to calculate when a specific dependency-management file (such as a `package.json` or `requirements.txt` in this case) changes so the cache will be updated accordingly. Also note that the `restore_cache` example uses interpolation to put dynamic values into the cache-key, allowing more control in what exactly constitutes the need to update a cache.

**Note:** Before adding any caching steps to your workflow, verify the dependencies installation step succeeds. Caching a failed dependency step will require you to change the cache key in order to avoid failed builds due to a bad cache. 

Because caching is a such a critical aspect of optimizing builds and workflows, you should first familiarize yourself with the following page that describes caching and how various strategies can be optimized:

- Caching - https://circleci.com/docs/2.0/caching/

## Improving Test Performance

> When running tests on the CircleCI platform, one of the primary considerations you will want to make is how you can optimize the testing process to minimize credit usage and improve overall testing performance and results. Testing can sometimes be a time and performance-intensive process; therefore, the ability to reduce testing time can be a significant boost to your organizational goals.

Reviewer 4: ```suggestion
When running tests on the CircleCI platform, one of the primary considerations you will want to make is how you can optimize the testing process to minimize credit usage and improve overall testing performance and results. Testing can sometimes be a time and performance-intensive process, therefore, the ability to reduce testing time can be a significant boost to your organizational goals.
```

There are many different test suites and approaches you can use when testing on the CircleCI platform. Although CircleCI is test suite agnostic, the example below (adapted with permission from the developer who wrote about this test optimization use case in his [blog post](https://www.brautaset.org/articles/2019/speed-up-circleci.html)) describes how you can optimize testing using Django and the CircleCI platform.

### Testing Optmization on the CircleCI Platform For a Python Django Project

Some organizations use CircleCI to run tests for each change before merging to the main branch. Faster tests means faster feedback cycles, which in turn means you can confidently ship code more often. Let's take a look at a case study for a Python Django application's workflow, that took more than 13 minutes to complete testing on the CircleCI platform.

An example of what the testing process might look like is shown below.

> [Test Optimization Process Before Optimization]({{site.baseurl}}/assets/img/docs/optimization_cookbook_workflow_optimization_1.png)

Reviewer 4: ```suggestion
![Test Optimization Process Before Optimization]({{site.baseurl}}/assets/img/docs/optimization_cookbook_workflow_optimization_1.png)
```

Let's take a closer look at the testing process in the figure above to better understand the time it took to complete the tests.

The following steps were performed during testing:

1) The build job created a Docker image, which contained only runtime dependencies. 
2) The build job dumped the image to a file with `docker save`, and then persisted it in the workspace. 

> 3) Two tests jobs were run to restore the base image from the workspace.

Reviewer 4: ```suggestion
3) Two test jobs were run to restore the base image from the workspace.
```

- Issuing `docker save` to dump the built image to a file took around **30** seconds.
- Persisting the image to the workspace added another **60** seconds.
- The test jobs then had to attach the workspace and load the base image, adding another **30** seconds to the testing.
- The test jobs started dependencies (Redis, Cassandra, and PostgreSQL) with `docker-compose`, which required the use of the machine executor. This added an additional **30-60** second startup time compared to the docker executors.
- Because the base image from the build job contained only runtime dependencies, a docker image had to be built, extending the base to add dependencies for testing. This added  about **70** seconds.

As you can see, there is a a significant amount of time being spent setting up the tests, without any actual tests being performed. In fact, this approach required 6.5 minutes before the actual tests were run, which took another 6.5 minutes.

### Test Preparation Optimization

> Knowing that ~13 minutes was too long to perform the steps in this workflow, the following approaches were taken to optimize and reduce the time it took to run this workflow.

Reviewer 4: ```suggestion
Knowing that ~13 minutes was too long to perform the steps in this workflow, the following approaches were taken to optimize and reduce this time.
```

> Knowing that ~13 minutes was too long to perform the steps in this workflow, the following approaches were taken to optimize this workflow to reduce the time it took to run this workflow.

Reviewer 4: ```suggestion
Knowing that ~13 minutes was too long to perform the steps in this workflow, the following approaches were taken to optimize and reduce the time it took to run this workflow.
```

#### Dependency Changes

Installing dependencies in the primary container on CircleCI, rather than relying on a Dockerfile, may enable you to use CircleCI's caching to speed up `virtualenv` creation.

### Test Execution Optimization

Now that the test preparation time has been reduced, you may also wish to speed up the running of the actual tests. For example, you may not need to keep the database after test runs. One way you could speed up testing is to replace the database image used for tests with an in-memory Postgres image that does not save to disk. Another method you may wish to take is to run your tests in parallel instead of one-test-at-a-time. 

The figure below illustrates how overall these changes can reduce the total workflow time.

> [Test Optimization Process After Optimization]({{site.baseurl}}/assets/img/docs/optimization_cookbook_workflow_optimization_2.png)

Reviewer 4: ```suggestion
![Test Optimization Process After Optimization]({{site.baseurl}}/assets/img/docs/optimization_cookbook_workflow_optimization_2.png)
```
